src/client/oicp/CpoOICPClient.ts

```diff
@@ -0,0 +1,1272 @@
+    const identification: OICPIdentification = {} as OICPIdentification;
+    const rfidIdentification: OICPRFIDmifarefamilyIdentification = {} as OICPRFIDmifarefamilyIdentification;
+    rfidIdentification.UID = transaction.tagID;
+    identification.RFIDMifareFamilyIdentification = rfidIdentification;
+
+    let siteArea;
+    if (!chargingStation.siteArea) {
+      siteArea = await SiteAreaStorage.getSiteArea(this.tenant.id, chargingStation.siteAreaID);
+    } else {
+      siteArea = chargingStation.siteArea;
+    }
+
+    const options = {
+      countryID: this.getLocalCountryCode(ServerAction.OICP_PUSH_SESSIONS),
+      partyID: this.getLocalPartyID(ServerAction.OICP_PUSH_SESSIONS),
+      addChargeBoxID: true
+    };
+    const oicpEvse = OICPMapping.getEvseByConnectorId(this.tenant, siteArea, chargingStation, transaction.connectorId, options);
+
+    const oicpSession: OICPSession = {} as OICPSession;
+    oicpSession.start_datetime = transaction.timestamp;
+    oicpSession.kwh = 0;
+    oicpSession.identification = identification;
+    oicpSession.evse = oicpEvse;
+    oicpSession.currency = this.settings.currency;
+    oicpSession.status = OICPSessionStatus.PENDING;
+    oicpSession.total_cost = transaction.currentCumulatedPrice > 0 ? transaction.currentCumulatedPrice : 0;
+    oicpSession.last_updated = transaction.timestamp;
+    transaction.oicpData = {
+      session: oicpSession
+    };
+
+    // Log
+    Logging.logDebug({
+      tenantID: this.tenant.id,
+      action: ServerAction.OICP_PUSH_SESSIONS,
+      message: `Start OICP Transaction ID (ID '${transaction.id}') to Hubject`,
+      module: MODULE_NAME, method: 'startSession',
+      detailedMessages: { payload: oicpSession }
+    });
+
+    // Call Hubject
+    const response = await this.authorizeStart(transaction);
+    // Save Session ID that was created by Hubject
+    transaction.oicpData.session.id = response.SessionID;
+    transaction.oicpData.session.empPartnerSessionID = response.EMPPartnerSessionID;
+    transaction.oicpData.session.providerID = response.ProviderID;
+
+    Logging.logDebug({
+      tenantID: this.tenant.id,
+      action: ServerAction.OICP_PUSH_SESSIONS,
+      message: `Start OICP Transaction ID '${transaction.oicpData.session.id}' (ID '${transaction.id}') response received from Hubject`,
+      module: MODULE_NAME, method: 'startSession',
+      detailedMessages: { response: response }
+    });
+  }
+
+  async updateSession(transaction: Transaction): Promise<void> {
+    if (!transaction.oicpData || !transaction.oicpData.session) {
+      throw new BackendError({
+        source: transaction.chargeBoxID,
+        action: ServerAction.OICP_PUSH_SESSIONS,
+        message: 'OICP Session not started',
+        module: MODULE_NAME, method: 'updateSession',
+      });
+    }
+
+    // Hubject restriction: "Progress Notification can be sent only at interval of at least 300 seconds." (5 Minutes)
```

Comment:
The way you implement that check force session update skipping. 
You should:
+ log if that happens;
+ identify the call trace in the log to avoid the caller use it so often.

Reply:
We should talk about that. The Progress notification is called with every new meter values from the charging station.

Reply:
The logging is still valid to identify the problematic caller. The caller can implement a duration counter a unix timestamp or similar format that cumulate the time between each meter values received for a transaction and call the function only when the counter is up to > 300 s. And then reset the counter. 

And also, do not use magic number in the code: 300, 432, etc. It's not maintainable, always define a constants or a data structure to store them with explicit name. That way our code will be more readable and maintainable by others.

Reply:
I changed it. The OICP session update that is part of the transaction is not skipped anymore.
The Progress update post to Hubject is send every 300 seconds. I created a constant for that in the Constants class and a method checkProgressUpdateInterval() that checks the difference between the now and the last progress update post to hubject. 
So I can ensure that Progress Notification  are sent only at interval of at least 300 seconds.